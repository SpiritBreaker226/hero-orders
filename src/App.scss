@import 'index';

.App-logo {
  height: 3.8em;
  pointer-events: none;
  background-color: $base-color;
}
