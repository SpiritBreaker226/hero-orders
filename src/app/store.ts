import { configureStore, ThunkAction, Action } from '@reduxjs/toolkit'
import tableReducer from '../features/table/tableSlice'
import fetchDataReducer from '../features/fetch_data/fetchDataSlice'
import searchWithVendorNameDataReducer from '../features/search_with_vendor_name/searchWithVendorNameSlice'

export const store = configureStore({
  reducer: {
    table: tableReducer,
    fetch: fetchDataReducer,
    searchWithVendorName: searchWithVendorNameDataReducer,
  },
})

export type RootState = ReturnType<typeof store.getState>
export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  Action<string>
>
