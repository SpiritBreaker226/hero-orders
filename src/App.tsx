import React from 'react'

import ErrorBoundary from './features/ErrorBoundary'
import FetchData from './features/fetch_data/FetchData'
import Table from './features/table/Table'

import logo from './logo.svg'
import './App.scss'

function App() {
  return (
    <ErrorBoundary>
      <header className="App-header">
        <img src={logo} className="App-logo" alt="chef hero" />
      </header>

      <main role="main">
        <Table />
        <FetchData />
      </main>
    </ErrorBoundary>
  )
}

export default App
