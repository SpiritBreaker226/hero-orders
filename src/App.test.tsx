import React from 'react'

import { Provider } from 'react-redux'

import { waitFor, screen, render, fireEvent } from '@testing-library/react'

import axios from 'axios'

import orders from './features/helpers/test_data_orders.json'

import { store } from './app/store'

import App from './App'

jest.mock('axios')
const mockedAxios = axios as jest.Mocked<typeof axios>

describe('App', () => {
  describe('when fetching order data', () => {
    beforeEach(async () => {
      mockedAxios.get.mockResolvedValue({
        data: orders,
      })

      render(
        <Provider store={store}>
          <App />
        </Provider>
      )

      await waitFor(() => screen.getByTestId('orders-table'))
    })

    it('renders orders component', () => {
      expect(mockedAxios.get).toHaveBeenCalledTimes(1)
      expect(mockedAxios.get).toHaveBeenCalledWith(
        process.env.REACT_APP_BASE_API_URL
      )
      expect(screen.getByRole('img')).toBeVisible()
      expect(screen.getByRole('main')).toMatchSnapshot()
    })

    describe('for searching', () => {
      describe('when selecting a vendor ', () => {
        it('should render only that vender then reset back', async () => {
          fireEvent.change(screen.getByTestId('searchWithVendorName'), {
            target: { value: 'D & D Poultry Ltd.' },
          })

          expect(
            screen.getAllByTestId('SupplierD & D Poultry Ltd.').length
          ).toEqual(2)
          expect(
            screen.queryByTestId('SupplierCanway Distribution')
          ).not.toBeInTheDocument()

          fireEvent.click(screen.getByTestId('searchResetButton'))

          expect(
            screen.getAllByTestId('SupplierD & D Poultry Ltd.').length
          ).toEqual(2)
          expect(
            screen.getAllByTestId('SupplierCanway Distribution').length
          ).toEqual(1)
        })
      })
    })
  })

  describe('when loading the page', () => {
    it('should not render no orders message and display loading text', async () => {
      mockedAxios.get.mockResolvedValueOnce({
        data: { data: [] },
      })

      render(
        <Provider store={store}>
          <App />
        </Provider>
      )

      expect(
        screen.queryByText('Loading', { exact: false })
      ).toBeInTheDocument()

      expect(
        screen.queryByText('No Orders Found', { exact: false })
      ).not.toBeInTheDocument()

      await waitFor(() => screen.getByTestId('no-orders'))

      expect(
        screen.queryByText('Loading', { exact: false })
      ).not.toBeInTheDocument()

      expect(screen.queryByTestId('searchBox')).not.toBeInTheDocument()

      expect(
        screen.queryByText('No Orders Found', { exact: false })
      ).toBeInTheDocument()
    })

    describe('on error from the server', () => {
      it('should render an error message', async () => {
        mockedAxios.get.mockRejectedValue(new Error('fake error message'))

        render(
          <Provider store={store}>
            <App />
          </Provider>
        )

        await waitFor(() => screen.getByText('error', { exact: false }))

        expect(
          screen.queryByText('No Orders Found', { exact: false })
        ).not.toBeInTheDocument()

        expect(
          screen.getByText('fake error message', { exact: false })
        ).toBeInTheDocument()
      })
    })
  })
})
