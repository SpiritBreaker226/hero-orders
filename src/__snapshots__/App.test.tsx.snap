// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App when fetching order data renders orders component 1`] = `
<main
  role="main"
>
  <section
    data-testid="orders-table"
  >
    <header>
      <h4>
        Status
      </h4>
      <h4>
        Delivery Day
      </h4>
      <h4>
        Supplier
      </h4>
      <h4>
        Total
      </h4>
    </header>
    <section
      role="row"
    >
      <span>
        <section
          aria-hidden="true"
        >
          <h4>
            Status
          </h4>
          <div>
            Paid
          </div>
        </section>
        <div>
          Paid
        </div>
      </span>
      <span>
        <section
          aria-hidden="true"
        >
          <h4>
            Delivery Day
          </h4>
          <div>
            Aug 7, 2020
          </div>
        </section>
        <div>
          Aug 7, 2020
        </div>
      </span>
      <span>
        <section
          aria-hidden="true"
        >
          <h4>
            Supplier
          </h4>
          <div>
            D & D Poultry Ltd.
            <span
              class="table-field-pending-vendor-onboarding"
            >
              1st
            </span>
          </div>
        </section>
        <div>
          D & D Poultry Ltd.
          <span
            class="table-field-pending-vendor-onboarding"
          >
            1st
          </span>
        </div>
      </span>
      <span>
        <section
          aria-hidden="true"
        >
          <h4>
            Total
          </h4>
          <div>
            $273.95
          </div>
        </section>
        <div>
          $273.95
        </div>
      </span>
    </section>
    <section
      role="row"
    >
      <span>
        <section
          aria-hidden="true"
        >
          <h4>
            Status
          </h4>
          <div>
            Delivered
          </div>
        </section>
        <div>
          Delivered
        </div>
      </span>
      <span>
        <section
          aria-hidden="true"
        >
          <h4>
            Delivery Day
          </h4>
          <div>
            Jul 31, 2020
          </div>
        </section>
        <div>
          Jul 31, 2020
        </div>
      </span>
      <span>
        <section
          aria-hidden="true"
        >
          <h4>
            Supplier
          </h4>
          <div>
            D & D Poultry Ltd.
            <span
              class="table-byos"
            >
              Market
            </span>
          </div>
        </section>
        <div>
          D & D Poultry Ltd.
          <span
            class="table-byos"
          >
            Market
          </span>
        </div>
      </span>
      <span>
        <section
          aria-hidden="true"
        >
          <h4>
            Total
          </h4>
          <div>
            $273.95
          </div>
        </section>
        <div>
          $273.95
        </div>
      </span>
    </section>
    <section
      role="row"
    >
      <span>
        <section
          aria-hidden="true"
        >
          <h4>
            Status
          </h4>
          <div>
            Paid
          </div>
        </section>
        <div>
          Paid
        </div>
      </span>
      <span>
        <section
          aria-hidden="true"
        >
          <h4>
            Delivery Day
          </h4>
          <div>
            Jul 31, 2020
          </div>
        </section>
        <div>
          Jul 31, 2020
        </div>
      </span>
      <span>
        <section
          aria-hidden="true"
        >
          <h4>
            Supplier
          </h4>
          <div>
            D & D Poultry Ltd.
            <span
              class="table-byos"
            >
              Market
            </span>
            <span
              class="table-field-pending-vendor-onboarding"
            >
              1st
            </span>
          </div>
        </section>
        <div>
          D & D Poultry Ltd.
          <span
            class="table-byos"
          >
            Market
          </span>
          <span
            class="table-field-pending-vendor-onboarding"
          >
            1st
          </span>
        </div>
      </span>
      <span>
        <section
          aria-hidden="true"
        >
          <h4>
            Total
          </h4>
          <div>
            $821.85
          </div>
        </section>
        <div>
          $821.85
        </div>
      </span>
    </section>
    <section
      role="row"
    >
      <span>
        <section
          aria-hidden="true"
        >
          <h4>
            Status
          </h4>
          <div>
            Paid
          </div>
        </section>
        <div>
          Paid
        </div>
      </span>
      <span>
        <section
          aria-hidden="true"
        >
          <h4>
            Delivery Day
          </h4>
          <div>
            Jul 31, 2020
          </div>
        </section>
        <div>
          Jul 31, 2020
        </div>
      </span>
      <span>
        <section
          aria-hidden="true"
        >
          <h4>
            Supplier
          </h4>
          <div>
            D & D Poultry Ltd.
          </div>
        </section>
        <div>
          D & D Poultry Ltd.
        </div>
      </span>
      <span>
        <section
          aria-hidden="true"
        >
          <h4>
            Total
          </h4>
          <div>
            $397.96
          </div>
        </section>
        <div>
          $397.96
        </div>
      </span>
    </section>
  </section>
</main>
`;
